### PR TITLE
Test DevitoBench

### PR DESCRIPTION
**Pull request by georgebisbas**
WIP-testing
Testing basic execution of cross product for modes and problems on DevitoBench.

### CODE REVIEW

File: requirements.txt [hunk 1]
```diff
@@ -21,3 +21,4 @@ frozendict
 anytree>=2.4.3
 pyrevolve>=2.1.3
 distributed>=1.27
+git+https://github.com/devitocodes/devitobench
```
Comment (rhodrin) [inline requirements.txt]: I don't think DevitoBench should be a default requirement.

Comment (georgebisbas) [reply requirements.txt]: It will probably not be. At the moment is just WIP-testing.
Related discussion happened under here: 
https://devitocodes.slack.com/archives/C7JMLMSG0/p1579101476448300

**Comment by mloubout** [comment]
Ok so I like how it's setup but I have two main concerns:
 - It is now just another test but quite costly so makes all the pytest a lot slower
 -  More than `isfile` would check if file is not empty maybe?

So maybe I would create a marker for `bench` or something so that can make a bench only test rather than add all of that to all the tests runs

File: requirements.txt (hunk 1 above)
Comment (FabioLuporini) [reply requirements.txt]: still WIP?

File: setup.py [hunk 2]
```diff
@@ -9,28 +9,24 @@
     optionals = f.read().splitlines()
 
 reqs = []
-links = []
 for ir in required:
     if ir[0:3] == 'git':
-        links += [ir + '#egg=' + ir.split('/')[-1] + '-0']
-        reqs += [ir.split('/')[-1]]
+        name = ir.split('/')[-1]
+        reqs += ['%s @ %s@master' % (name, ir)]
```
Comment (FabioLuporini) [inline setup.py]: @mloubout are all these changes sane?

File: tests/test_benchmark.py [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+    Test combinations of modes x problems and created .json filename.
+    Testing for both OpenMP enabled or disabled.
+    """
+    if configuration['openmp']:
+        nthreads = configuration['platform'].cores_physical
+    else:
+        nthreads = 1
+
+    command_bench = ['python', 'benchmarks/user/benchmark.py', mode, '-P', problem,
+                     '-d', '40', '40', '40', '--tn', '20', '-x', '1']
```
Comment (FabioLuporini) [inline tests/test_benchmark.py]: How about making it even smaller/shorter:
```
-d 16 16 16 
--tn 4 
```

File: tests/test_benchmark.py [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+    try:
+        check_call(command_bench)
+    except CalledProcessError as e:
+        raise CompilationError('Command "%s" return error status %d. '
```
Comment (FabioLuporini) [inline tests/test_benchmark.py]: why is this needed ?

File: tests/test_benchmark.py [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+                               'Unable to compile code.\n' %
```
Comment (FabioLuporini) [inline tests/test_benchmark.py]: and most importantly, what makes you think that's due to an error with compilation?

File: tests/test_benchmark.py [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+                               (e.cmd, e.returncode))
+    dir_name = 'results/'
+
+    base_filename = problem
```
Comment (FabioLuporini) [inline tests/test_benchmark.py]: **all** the underscores in the list of variables below should be dropped

then you just do:

```
bench_file = path.join('_'.join(....))
```

File: tests/test_benchmark.py [hunk 7]
```diff
@@ -0,0 +1,70 @@
...
+    to = '_to[2]'
+    dse = '_dse[advanced]'
+    dle = '_dle[advanced]'
+    at = '_at[aggressive]'
+    nt = path.join('_nt[' + str(nthreads) + ']')
+    mpi = '_mpi[False]'
+    np = '_np[1]'
+    rank = '_rank[0]'
+
+    listdir("results/")
```
Comment (FabioLuporini) [inline tests/test_benchmark.py]: ?

File: tests/test_benchmark.py [hunk 8]
```diff
@@ -0,0 +1,70 @@
...
+    bench_file = path.join(dir_name + base_filename + arch + shape + nbl + tn + so + to
+                           + dse + dle + at + nt + mpi + np + rank + filename_suffix)
+
+    command_plot = ['python', 'benchmarks/user/benchmark.py', 'plot', '-P', problem,
+                    '-d', '40', '40', '40', '--tn', '20', '--max-bw', '12.8',
+                    '--flop-ceil', '80', 'linpack']
+    try:
+        check_call(command_plot)
+    except CalledProcessError as e:
+        raise CompilationError('Command "%s" return error status %d. '
```
Comment (FabioLuporini) [inline tests/test_benchmark.py]: ?

File: tests/test_benchmark.py [hunk 9]
```diff
@@ -0,0 +1,70 @@
...
+    mpi = '_mpi[False]'
+    np = '_np[1]'
+    rank = '_rank[0]'
+
+    listdir("results/")
+
+    bench_file = path.join(dir_name + base_filename + arch + shape + nbl + tn + so + to
+                           + dse + dle + at + nt + mpi + np + rank + filename_suffix)
+
+    command_plot = ['python', 'benchmarks/user/benchmark.py', 'plot', '-P', problem,
```
Comment (FabioLuporini) [inline tests/test_benchmark.py]: plotting should be a separate test, if we really want to test it

File: tests/test_benchmark.py [hunk 10]
```diff
@@ -0,0 +1,70 @@
...
+
+
+pytestmark = skipif(['yask', 'ops'])
+
+
+@pytest.mark.parametrize('mode', ['bench'])
+@pytest.mark.parametrize('problem', ['acoustic', 'tti', 'elastic', 'viscoelastic'])
+def test_bench(mode, problem):
+    """
+    Test combinations of modes x problems and created .json filename.
```
Comment (FabioLuporini) [inline tests/test_benchmark.py]: x -> and

Drop "and created .json filename" (also there was a typo)

File: tests/test_benchmark.py [hunk 11]
```diff
@@ -0,0 +1,70 @@
...
+                    '-d', '40', '40', '40', '--tn', '20', '--max-bw', '12.8',
+                    '--flop-ceil', '80', 'linpack']
+    try:
+        check_call(command_plot)
+    except CalledProcessError as e:
+        raise CompilationError('Command "%s" return error status %d. '
+                               'Unable to compile code.\n' %
+                               (e.cmd, e.returncode))
+
+    bkend = '_bkend[core]'
```
Comment (FabioLuporini) [inline tests/test_benchmark.py]: drop underscore

File: tests/test_benchmark.py [hunk 12]
```diff
@@ -0,0 +1,70 @@
...
+
+    plot_file = path.join(dir_name + base_filename + shape + so + to + arch + bkend + at)
```
Comment (FabioLuporini) [inline tests/test_benchmark.py]: same story -- `'_'.join(...)`

File: setup.py (hunk 2 above)
Comment (mloubout) [reply setup.py]: Yes, I made these newer `setuptool` standard, actually works better than before.

File: requirements.txt (hunk 1 above)
Comment (georgebisbas) [reply requirements.txt]: just a bit more

File: tests/test_benchmark.py (hunk 10 above)
Comment (georgebisbas) [reply tests/test_benchmark.py]: Sure

File: tests/test_benchmark.py (hunk 3 above)
Comment (georgebisbas) [reply tests/test_benchmark.py]: Sure

File: tests/test_benchmark.py (hunk 5 above)
Comment (georgebisbas) [reply tests/test_benchmark.py]: They are just leftovers from a pytest template.

File: tests/test_benchmark.py (hunk 7 above)
Comment (georgebisbas) [reply tests/test_benchmark.py]: To check created files locally, will be dropped.

File: tests/test_benchmark.py (hunk 9 above)
Comment (georgebisbas) [reply tests/test_benchmark.py]: Can I ensure that two tests are excecuted sequentially? So that pytest doesn't execute plot before bench?

File: tests/test_benchmark.py (hunk 8 above)
Comment (georgebisbas) [reply tests/test_benchmark.py]: Same, to be dropped.

File: tests/test_benchmark.py (hunk 6 above)
Comment (georgebisbas) [reply tests/test_benchmark.py]: this is probably wrong? It creates this: `a_c_o_u_s_t_i_c_a_r_c_h_[_u_n_k_n_o_w_n_]_s_`

File: tests/test_benchmark.py (hunk 11 above)
Comment (georgebisbas) [reply tests/test_benchmark.py]: Same yes

File: tests/test_benchmark.py (hunk 6 above)
Comment (georgebisbas) [reply tests/test_benchmark.py]: ok, found the way

**Comment by FabioLuporini** [comment]
Can we update tags and title of this PR? are we close to landing ?

File: requirements.txt (hunk 1 above)
Comment (FabioLuporini) [reply requirements.txt]: this should be an optional dependence

if you use the mode `run`, instead of `bench`, you shouldn't need `devitobench` at all. That's what I would do, I think

File: tests/test_benchmark.py [hunk 13]
```diff
@@ -0,0 +1,61 @@
...
+    nt = path.join('nt[' + str(nthreads) + ']')
+    mpi = 'mpi[False]'
+    np = 'np[1]'
+    rank = 'rank[0]'
+    bkend = 'bkend[core]'
+
+    bench_corename = path.join('_'.join([base_filename, arch, shape, nbl, tn,
+                               so, to, dse, dle, at, nt, mpi, np, rank]))
+
+    bench_filename = path.join(dir_name + bench_corename + filename_suffix)
```
Comment (FabioLuporini) [inline tests/test_benchmark.py]: this should be:

```
path.join(dir_name, bench_corename, filename_suffix)
```

Otherwise it's totally useless to use `path.join`

Comment (FabioLuporini) [reply tests/test_benchmark.py]: ignore the comment above, I think? but why do you need path.join ?

File: tests/test_benchmark.py (hunk 9 above)
Comment (FabioLuporini) [reply tests/test_benchmark.py]: fair point. I don't think you can, AFAICT after 1 min of googling

File: tests/test_benchmark.py [hunk 14]
```diff
@@ -0,0 +1,61 @@
...
+    Testing for both OpenMP enabled or disabled.
+    """
+    if configuration['openmp']:
+        nthreads = configuration['platform'].cores_physical
+    else:
+        nthreads = 1
+
+    command_bench = ['python', 'benchmarks/user/benchmark.py', mode, '-P', problem,
+                     '-d', '16', '16', '16', '--tn', '4', '-x', '1']
+    check_call(command_bench)
```
Comment (FabioLuporini) [inline tests/test_benchmark.py]: did you verify that if you for example supply an unknown flag to the command , the test actually fails ?

For example:
```
['python', 'benchmarks/user/benchmark.py', mode, '-P', problem, ......, '-asdjoias', ...]
```

and, analogously, if there were an error in the execution of the benchmark, the test fails as well?

just to be sure...

File: tests/test_benchmark.py [hunk 15]
```diff
@@ -0,0 +1,61 @@
...
+    arch = 'arch[unknown]'
+    shape = 'shape[16,16,16]'
+    nbl = 'nbl[10]'
+    tn = 'tn[4]'
+    so = 'so[2]'
+    to = 'to[2]'
+    dse = 'dse[advanced]'
+    dle = 'dle[advanced]'
+    at = 'at[aggressive]'
+    nt = path.join('nt[' + str(nthreads) + ']')
```
Comment (FabioLuporini) [inline tests/test_benchmark.py]: in devito, we do **not** concatenate strings with '+'. We rather use

```
"nt[%d]" % nthreads
```

File: tests/test_benchmark.py [hunk 16]
```diff
@@ -0,0 +1,61 @@
...
+    bench_filename = path.join(dir_name + bench_corename + filename_suffix)
+
+    command_plot = ['python', 'benchmarks/user/benchmark.py', 'plot', '-P', problem,
+                    '-d', '16', '16', '16', '--tn', '4', '--max-bw', '12.8',
+                    '--flop-ceil', '80', 'linpack']
+    check_call(command_plot)
+
+    plot_corename = path.join('_'.join([base_filename, shape, so, to, arch,
+                              bkend, at]))
+    plot_filename = path.join(dir_name + plot_corename)
```
Comment (FabioLuporini) [inline tests/test_benchmark.py]: no string concatenation with +

File: tests/test_benchmark.py [hunk 17]
```diff
@@ -0,0 +1,61 @@
...
+    Test combinations of modes and problems.
+    Testing for both OpenMP enabled or disabled.
+    """
+    if configuration['openmp']:
+        nthreads = configuration['platform'].cores_physical
+    else:
+        nthreads = 1
+
+    command_bench = ['python', 'benchmarks/user/benchmark.py', mode, '-P', problem,
+                     '-d', '16', '16', '16', '--tn', '4', '-x', '1']
```
Comment (FabioLuporini) [inline tests/test_benchmark.py]: out of curiosity, how many timesteps does '--tn 4' imply?

File: tests/test_benchmark.py [hunk 18]
```diff
@@ -0,0 +1,61 @@
...
+
+
+pytestmark = skipif(['yask', 'ops'])
+
+
+@pytest.mark.parametrize('mode', ['bench'])
+@pytest.mark.parametrize('problem', ['acoustic', 'tti', 'elastic', 'viscoelastic'])
+def test_bench(mode, problem):
+    """
+    Test combinations of modes and problems.
```
Comment (FabioLuporini) [inline tests/test_benchmark.py]: "Test the Devito benchmark framework on various combinations of modes and problems."

File: tests/test_benchmark.py [hunk 19]
```diff
@@ -0,0 +1,61 @@
...
+    Testing for both OpenMP enabled or disabled.
```
Comment (FabioLuporini) [inline tests/test_benchmark.py]: Drop, I'd say irrelevant/obvious.

Also, it should have read "Test both ..."

File: tests/test_benchmark.py [hunk 20]
```diff
@@ -0,0 +1,61 @@
...
+    command_plot = ['python', 'benchmarks/user/benchmark.py', 'plot', '-P', problem,
+                    '-d', '16', '16', '16', '--tn', '4', '--max-bw', '12.8',
+                    '--flop-ceil', '80', 'linpack']
+    check_call(command_plot)
+
+    plot_corename = path.join('_'.join([base_filename, shape, so, to, arch,
+                              bkend, at]))
+    plot_filename = path.join(dir_name + plot_corename)
+
+    assert path.isfile(bench_filename)
```
Comment (FabioLuporini) [inline tests/test_benchmark.py]: This assert must be moved somewhere around line 50

File: tests/test_benchmark.py (hunk 14 above)
Comment (georgebisbas) [reply tests/test_benchmark.py]: Yes, it does fail, I have tested mispellings, wrong args/unkonwn flags, Well this how we realized the error that you fixed in https://github.com/devitocodes/devito/pull/1080

File: tests/test_benchmark.py (hunk 15 above)
Comment (georgebisbas) [reply tests/test_benchmark.py]: Hell, sorry , that was very stupid.

File: tests/test_benchmark.py (hunk 13 above)
Comment (georgebisbas) [reply tests/test_benchmark.py]: Well, bad leftovers...

File: tests/test_benchmark.py (hunk 9 above)
Comment (georgebisbas) [reply tests/test_benchmark.py]: hmm... lets see.

File: tests/test_benchmark.py (hunk 16 above)
Comment (georgebisbas) [reply tests/test_benchmark.py]: Moved to "%s%s"

File: tests/test_benchmark.py (hunk 20 above)
Comment (georgebisbas) [reply tests/test_benchmark.py]: Moved, fixed.

**Comment by georgebisbas** [comment]
> How long does this test take **overall** ?

90 secs sequentially locally

==================================================== slowest 20 test durations =====================================================
37.90s call     tests/test_benchmark.py::test_bench[tti-bench]
24.52s call     tests/test_benchmark.py::test_bench[viscoelastic-bench]
14.03s call     tests/test_benchmark.py::test_bench[elastic-bench]
8.01s call     tests/test_benchmark.py::test_bench[acoustic-bench]

**Comment by mloubout** [comment]
>90 secs sequentially locally

@FabioLuporini @georgebisbas there is three of it in the 20 slowest in the log for ~2min so I'd say overall adds about 3min to each run, and that's with tn=4, so maybe better to have its own build (like osx only runs one)

**Comment by mloubout** [comment]
Looks like something went wrong with rebase

**Comment by georgebisbas** [comment]
> Looks like something went wrong with rebase

Where can you see this? As I can see it is rebased on latest master.

**Comment by georgebisbas** [comment]
> > Looks like something went wrong with rebase
> 
> Where can you see this? As I can see it is rebased on latest master.

Sorry, I mean what is wrong?

File: requirements.txt (hunk 1 above)
Comment (FabioLuporini) [reply requirements.txt]: nonetheless, it should be an optional dependence, so be moved to a different file (@mloubout ?)

File: tests/test_benchmark.py (hunk 14 above)
Comment (FabioLuporini) [reply tests/test_benchmark.py]: good

File: tests/test_benchmark.py [hunk 21]
```diff
@@ -0,0 +1,60 @@
...
+
+    dir_name = 'results/'
+
+    base_filename = problem
+    filename_suffix = '.json'
+    arch = 'arch[unknown]'
+    shape = 'shape[16,16,16]'
+    nbl = 'nbl[10]'
+    tn = 'tn[4]'
+    so = 'so[2]'
```
Comment (FabioLuporini) [inline tests/test_benchmark.py]: @mloubout `so[2]` and TTI running? 😮

File: requirements.txt (hunk 1 above)
Comment (mloubout) [reply requirements.txt]: Still not sure what to do with these optionals requirements

File: tests/test_benchmark.py (hunk 21 above)
Comment (mloubout) [reply tests/test_benchmark.py]: Yeah it runs, not recommended for the quality of solution but runs fine

**Comment by FabioLuporini** [comment]
> @FabioLuporini @georgebisbas there is three of it in the 20 slowest in the log for ~2min so I'd say overall adds about 3min to each run, and that's with tn=4, so maybe better to have its own build (like osx only runs one)

but @mloubout , what was the runtime of 
https://github.com/devitocodes/devito/blob/master/azure-pipelines.py#L30

Perhaps what we should consider is the difference? ie how much slow*er* are we now ?

I don't know if I like the idea of moving it do a different action

**Comment by mloubout** [comment]
> what was the runtime

pipeline is running all examples too, but takes about 25-30min

>I don't know if I like the idea of moving it do a different action

I'm not saying that I'm just saying to have one of the pytests build to do just this one.

File: requirements.txt (hunk 1 above)
[the remark this answers is not in the document]
Comment (mloubout) [reply requirements.txt]: But then need to install MPI because optionals have mpi4py
